### ISSUE TITLE
Videos: Support for HTTP Live Streaming (HLS) with M3U8

### ISSUE THREAD

**Issue by tezza1971** [photoprism/photoprism#4179]
I did a search, and I haven't seen this yet, but I think HLS support would be idea. I understand that photoprism is not meant to be a video hosting platform but even short clips take a long time to start playing because it seems the whole thing needs to download before playback begins. It makes photoprism look like it's slow when it is far from slow for everything else.

I've been using HLS with my PeerTube instance and it really makes the platform feel snappy and a joy to use. Thats the way things are with PhotoPrism on photos so it makes sense to use HLS because it's the low hanging fruit to get the video playback experience up to speed. 

I did a quick and dirty inquiry using AI and HLS doesn't look overly difficult to implement using Video.js

https://www.perplexity.ai/search/How-hard-is-rBxuRFIOREGSJMae.6euSA

I'm not sure but I think even without multiple bitrate versions of the video clip, the chunking down of the big video blob makes it scrubbable before total download - which I think is a big deal.

What do you think?

**Comment by lastzero** [comment]
> It seems the whole thing needs to download before playback begins

This is not true. Browsers usually perform so-called "range requests" for this and PhotoPrism supports that. However, streaming only starts after the entire video has been transcoded (if needed), as the required metadata is gathered during transcoding and then (in the last step!) placed at the beginning of the transcoded video file:
- https://trac.ffmpeg.org/wiki/Encode/H.264#faststartforwebvideo
- https://docs.photoprism.app/getting-started/advanced/transcoding/

**Note that in some cases, e.g. with files created by other software, this required metadata MAY be at the end of the file and not at the beginning, so - as described in the documentation linked above - the video must be downloaded completely before it can be played by the browser.**

I understand that we could split long videos into smaller files and assemble them into a full length video with an M3U8 playlist, so playback can start while some files are still being transcoded (and possibly different bitrates could be offered).

However, the videos would still need to be transcoded and users might then not have a single regular AVC video file (with [Faststart](https://trac.ffmpeg.org/wiki/Encode/H.264#faststartforwebvideo)) in their library, but a bunch of files that they don't know how to play e.g. with Windows Explorer:
- https://en.wikipedia.org/wiki/M3U

There are pros and cons, but we'd be happy to look at it and also accept well-tested pull requests (there are actually quite a few Go libraries on GitHub that we might be able to integrate or use as inspiration if their license permits):
- https://docs.photoprism.app/developer-guide/pull-requests/
- https://github.com/m1k1o/go-transcode
- https://github.com/Roverr/rtsp-stream

**Comment by ChR-iSz** [comment]
Hello,

There are various ways to play (stream) the videos without having to load or buffer the entire video.
In general, I would refrain from transcoding if the video is already available in the correct mp4 (h264) format. Transcoding would be unnecessary and a waste of resources.

To put the required MP4 metadata (moov atom) at the beginning (which can indeed be at the end of the file, qt-faststart helps)
Example:
`qt-faststart input.mp4 output.mp4`

The process runs extremely fast (need r.b. 2 seconds for a 90 min video). It will only move the moov atom information to the beginning of the file.

[qt-faststart](https://manpages.debian.org/bookworm/ffmpeg/qt-faststart.1.en.html)

**Comment by lastzero** [comment]
PhotoPrism doesn't transcode MPEG4 AVC / H.264 as pretty much all browsers can play it natively. If the browser supports it and the bitrate is below the configured limit, HEVC / H.265 can also be played without transcoding. When transcoding videos, the faststart option is used by default. However, it is not applied to any existing MP4 / MOV files so you would need to do that manually. I wasn't aware that there is an extra  "qt-faststart" command for this. Thanks for pointing that out!
